File: src/main/java/com/raoulvdberge/refinedstorage/inventory/fluid/FilterIconFluidInventory.java
package com.raoulvdberge.refinedstorage.inventory.fluid;

import com.raoulvdberge.refinedstorage.item.FilterItem;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.CompoundNBT;
import net.minecraftforge.fluids.FluidStack;

public class FilterIconFluidInventory extends FluidInventory {
    public FilterIconFluidInventory(ItemStack stack) {
        super(1, Integer.MAX_VALUE);

        this.addListener((handler, slot, reading) -> {
            if (!stack.hasTag()) {
                stack.setTag(new CompoundNBT());
            }

            FilterItem.setFluidIcon(stack, getFluid(slot));
        });

        FluidStack icon = FilterItem.getFluidIcon(stack);
        if (icon != null) {
            setFluid(0, icon);
        }
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/inventory/fluid/FilterIconFluidInventory.kt
package com.refinedmods.refinedstorage.inventory.fluid

import com.refinedmods.refinedstorage.item.FilterItem
import net.minecraft.item.ItemStack
import net.minecraft.nbt.CompoundTag
import net.minecraftforge.fluids.FluidInstance

class FilterIconFluidInventory(stack: ItemStack) : FluidInventory(1, Int.MAX_VALUE) {
    init {
        addListener { handler: FluidInventory?, slot: Int, reading: Boolean ->
            if (!stack.hasTag()) {
                stack.tag = CompoundTag()
            }
            FilterItem.setFluidIcon(stack, getFluid(slot))
        }
        val icon: FluidInstance = FilterItem.getFluidIcon(stack)
        if (!icon.isEmpty()) {
            setFluid(0, icon)
        }
    }
}